src/rust/engine/fs/src/directory.rs

```diff
@@ -650,7 +650,7 @@ impl DigestTrie {
                     link_depth += 1;
                     if s.target == Component::CurDir.as_os_str() {
                         if link_depth >= MAX_LINK_DEPTH {
-                            warn!("Exceeded the maximum link depth while traversing links. Stopping traversal.");
+                            warn!("Exceeded the maximum link depth while traversing link `{}` to path {:#?}. Stopping traversal.", s.name, s.target);
```

Comment:
Minor change, but I think `s.name` is just the tail end of the path, so potentially misses some context, as someone might end up with a a message like `Exceeded maximum link depth while traversing link `foo` to path "..."` and maybe has many links called `foo`.

We could potentially print it like:

```suggestion
                            warn!("Exceeded the maximum link depth while traversing link {:#?} to path {:#?}. Stopping traversal.", path_so_far, s.target);
```

Which'll include more context automatically.

Reply:
> Actually looking at this, it only fires if the target == "." which guarantees an infinite symlink - do we even need to check link depth in this case? We're just going to end up traversing the same directory again and again

Good catch! I agree that if it's literally a link to `.` then warning immediately seems fine. A quick skim of the callers of `walk` suggest that most of them use it to collect a list of files or directories, and having duplicates of `self/self/self/.../file.txt` doesn't seem amazingly helpful.

Maybe @thejcannon should confirm, though.

> Actually could we even go a step further and detect the infinite symlink before we jump into it at all? It looks like Rust doesn't have a standard library crate for calculating relative paths but there are mature crates out there that do that.

I don't think we can detect all cases without traversing the file system, only "obvious" ones like linking to `.`, `..` or any parent.

That said... I'm personally inclined too not spend too much energy on this: having infinite symlink cycles seems like it's likely to be a mistake in most cases (?), so I think our goal should be to make sure the diagnostics are good enough to help the user fix them, so others don't suffer through the confusion you did. (Rather than optimising the detection completely.)

Reply:
So I took a stab at simplifying this a little without going to overboard, mostly for my own knowledge but I think I'm a little confused on how this function is working in main now. Namely, when the function discovers a symlink that references `.` it increments the link depth, calls `walk_helper` on the same path, and then returns. This guarantees nothing past this symlink in the `DigestTrie` is ever going to get walked, does it not?

Also, one other minor thing I noticed is that `link_depth` is incremented for every symlink discovered, rather than when we jump into a directory from a symlink. So in theory you could have 64 perfectly fine symlinks and it would complain about link depth.